src/background/extension.js

```diff
@@ -94,42 +57,105 @@ export class Extension {
       state.activateTab(/** @type {number} */ (tab.id));
     };
 
-    async function restoreSavedTabState() {
+    /**
+     * Initialize cached state for browser tabs by querying each tab to see
+     * whether the extension is active there.
+     *
+     * This should be called when the extension is loaded or reloaded.
+     */
+    const initTabStates = async () => {
       const tabs = await chromeAPI.tabs.query({});
-      const tabIds = tabs
-        .filter(tab => tab.id !== undefined)
-        .map(({ id }) => /** @type {number} */ (id));
-      store.reload(tabIds);
-      state.load(store.all());
-      tabIds.forEach(tabId => {
-        onTabStateChange(tabId, state.getState(tabId));
-      });
-    }
+      const activeStates = await Promise.all(
+        tabs.map(async tab => {
+          if (tab.id === undefined) {
+            return false;
+          }
+          try {
+            const active = await sidebar.isClientActiveInTab(tab);
+            return active;
+          } catch (e) {
+            console.warn(
+              `Unable to determine extension state in tab ${tab.id}`,
+              e
+            );
+            return false;
+          }
+        })
+      );
+
+      for (let i = 0; i < tabs.length; i++) {
+        const tab = tabs[i];
+        if (!tab.id) {
+          continue;
+        }
+        const isActive = activeStates[i];
+
+        // nb. If tab status is not available, we optimistically assume it is
+        // loaded.
+        const ready =
+          tab.status === 'complete' || typeof tab.status !== 'string';
+
+        state.setState(tab.id, {
+          state: isActive ? 'active' : 'inactive',
+          extensionSidebarInstalled: isActive,
+          ready,
+        });
```

Comment:
Something to note here is that the only per-tab state we populate after reloading the extension is:

1. Whether the client is active
2. Whether the tab is loaded

Additional state such as cached annotation counts (on the badge) and injection errors (when the badge displays `!` instead of a count) are lost. I think that's OK. The errors will appear again if you retry activating the client and the badge will update once you interact with the tab again.